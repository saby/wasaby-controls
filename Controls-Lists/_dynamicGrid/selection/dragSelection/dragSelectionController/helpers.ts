import { TPoint, TRect } from './types';
import { TResizerDirection } from '../shared/types';
import { START_FIXED_PART_OBSERVER, END_FIXED_PART_OBSERVER } from '../../../shared/constants';
import { TKeyPair } from './../../shared/types';
import * as React from 'react';

const PERCENT_OF_SIZE_TO_SELECT_CELL = 0.1;

export function getDragStartParams(
    container: HTMLElement,
    topLeftRect: TRect,
    bottomRightRect: TRect,
    direction: TResizerDirection
): {
    startPoint: TPoint;
    xOffset: number;
    yOffset: number;
    xMinOffset: number;
    yMinOffset: number;
} {
    const viewPortRect = getViewportBoundingClientRect(container);
    let xOffset: number;
    let yOffset: number;
    let xMinOffset: number;
    let yMinOffset: number;

    const startPoint: TPoint = {
        x: undefined,
        y: undefined,
    };

    const getXInViewportRange = (x: number) =>
        Math.min(Math.max(x, viewPortRect.x), viewPortRect.x + viewPortRect.width);
    const getYInViewportRange = (y: number) =>
        Math.min(Math.max(y, viewPortRect.y), viewPortRect.y + viewPortRect.height);

    const getXOffsetInViewportRange = (startX: number, offset: number) => {
        // Итоговая координата может быть уменьшена, если выделение вылезает за вьюпорт.
        const correctStartX = getXInViewportRange(startX + offset);
        // Реальный отступ, будет такой же или меньше.
        return offset > 0 ? correctStartX - startX : -1 * (startX - correctStartX);
    };

    const getYOffsetInViewportRange = (startY: number, offset: number) => {
        // Итоговая координата может быть уменьшена, если выделение вылезает за вьюпорт.
        const correctStartY = getYInViewportRange(startY + offset);
        // Реальный отступ, будет такой же или меньше.
        return offset > 0 ? correctStartY - startY : -1 * (startY - correctStartY);
    };

    // Координата Y
    // Начальная точка это либо левый верхний угол, либо правый нижний, но не больше чем размеры контейнера.
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'topRight':
            // Правый нижний, центр вокруг которого будет выделение.
            startPoint.y = getYInViewportRange(bottomRightRect.y + bottomRightRect.height);
            // Из левого верхнего вычитаем центр, чтобы узнать отступ(должен быть отрицательным).
            yOffset = getYOffsetInViewportRange(startPoint.y, topLeftRect.y - startPoint.y);
            break;
        case 'bottom':
        case 'right':
        case 'bottomLeft':
        case 'bottomRight':
            // Левый верхний угол, центр вокруг которого будет выделение.
            startPoint.y = getYInViewportRange(topLeftRect.y);
            // Из правого нижнего вычитаем центр, чтобы узнать отступ(должен быть положительным).
            yOffset = getYOffsetInViewportRange(
                startPoint.y,
                bottomRightRect.y + bottomRightRect.height - startPoint.y
            );
            break;
    }

    // Координата X
    // Начальная точка это либо левый верхний угол, либо правый нижний, но не больше чем размеры контейнера.
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            // Правый нижний, центр вокруг которого будет выделение.
            startPoint.x = getXInViewportRange(bottomRightRect.x + bottomRightRect.width);
            // Из левого верхнего вычитаем центр, чтобы узнать отступ(должен быть отрицательным).
            xOffset = getXOffsetInViewportRange(startPoint.x, topLeftRect.x - startPoint.x);
            break;
        case 'bottom':
        case 'right':
        case 'topRight':
        case 'bottomRight':
            // Левый верхний угол, центр вокруг которого будет выделение.
            startPoint.x = getXInViewportRange(topLeftRect.x);
            // Из правого нижнего вычитаем центр, чтобы узнать отступ(должен быть положительным).
            xOffset = getXOffsetInViewportRange(
                startPoint.x,
                bottomRightRect.x + bottomRightRect.width - startPoint.x
            );
            break;
    }

    // Минимальное смещение по Y
    switch (direction) {
        case 'top':
        case 'topLeft':
        case 'topRight':
            yMinOffset = -1 * bottomRightRect.height;
            break;
        case 'bottom':
        case 'bottomLeft':
        case 'bottomRight':
            yMinOffset = topLeftRect.height;
            break;
        case 'left':
        case 'right':
            yMinOffset = yOffset;
            break;
    }

    // Минимальное смещение по X
    switch (direction) {
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            xMinOffset = -1 * bottomRightRect.width;
            break;
        case 'right':
        case 'topRight':
        case 'bottomRight':
            xMinOffset = topLeftRect.width;
            break;
        case 'top':
        case 'bottom':
            xMinOffset = xOffset;
            break;
    }

    return {
        startPoint,
        xOffset,
        yOffset,
        xMinOffset,
        yMinOffset,
    };
}

export function getNewOffset(
    cellRect: TRect,
    startPoint: TPoint,
    direction: TResizerDirection,
    minXOffset: number,
    minYOffset: number
): {
    x: number;
    y: number;
} {
    let newOffsetX: number;
    let newOffsetY: number;

    // Координата Y
    switch (direction) {
        case 'top':
        case 'topLeft':
        case 'topRight':
            newOffsetY = cellRect.y - startPoint.y;
            newOffsetY = newOffsetY > 0 ? minYOffset : newOffsetY;
            break;
        case 'bottom':
        case 'bottomLeft':
        case 'bottomRight':
            newOffsetY = cellRect.y + cellRect.height - startPoint.y;
            newOffsetY = newOffsetY < 0 ? minYOffset : newOffsetY;
            break;
        case 'left':
        case 'right':
            newOffsetY = minYOffset;
            break;
    }

    // Координата X
    switch (direction) {
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            newOffsetX = cellRect.x - startPoint.x;
            newOffsetX = newOffsetX > 0 ? minXOffset : newOffsetX;
            break;
        case 'right':
        case 'topRight':
        case 'bottomRight':
            newOffsetX = cellRect.x + cellRect.width - startPoint.x;
            newOffsetX = newOffsetX < 0 ? minXOffset : newOffsetX;
            break;
        case 'top':
        case 'bottom':
            newOffsetX = minXOffset;
            break;
    }

    return {
        x: newOffsetX,
        y: newOffsetY,
    };
}

export function getEndPoint(
    startPoint: TPoint,
    direction: TResizerDirection,
    offsetX: number,
    offsetY: number
): TPoint {
    let x: number;
    let y: number;

    // Координата Y
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'topRight':
            y = startPoint.y - -1 * offsetY + 2;
            break;
        case 'bottom':
        case 'right':
        case 'bottomLeft':
        case 'bottomRight':
            y = startPoint.y + offsetY - 2;
            break;
    }

    // Координата X
    switch (direction) {
        case 'top':
        case 'left':
        case 'topLeft':
        case 'bottomLeft':
            x = startPoint.x - -1 * offsetX + 2;
            break;
        case 'bottom':
        case 'right':
        case 'topRight':
        case 'bottomRight':
            x = startPoint.x + offsetX - 2;
            break;
    }

    return {
        x,
        y,
    };
}

type TKeyPairs = {
    topLeft: TKeyPair;
    bottomRight: TKeyPair;
};

export function getDragEndParams(
    startKeyPairs: TKeyPairs,
    endKeyPair: TKeyPair,
    direction: TResizerDirection
): TKeyPairs {
    const result: TKeyPairs = {
        topLeft: {
            itemKey: undefined,
            columnKey: undefined,
        },
        bottomRight: {
            itemKey: undefined,
            columnKey: undefined,
        },
    };
    switch (direction) {
        case 'top': {
            result.topLeft.itemKey = endKeyPair.itemKey;
            result.topLeft.columnKey = startKeyPairs.topLeft.columnKey;
            result.bottomRight.itemKey = startKeyPairs.bottomRight.itemKey;
            result.bottomRight.columnKey = startKeyPairs.bottomRight.columnKey;
            break;
        }
        case 'topLeft': {
            result.topLeft = endKeyPair;
            result.bottomRight = startKeyPairs.bottomRight;
            break;
        }
        case 'topRight': {
            result.topLeft.itemKey = endKeyPair.itemKey;
            result.topLeft.columnKey = startKeyPairs.topLeft.columnKey;
            result.bottomRight.itemKey = startKeyPairs.bottomRight.itemKey;
            result.bottomRight.columnKey = endKeyPair.columnKey;
            break;
        }
        case 'bottom': {
            result.topLeft.itemKey = startKeyPairs.topLeft.itemKey;
            result.topLeft.columnKey = startKeyPairs.topLeft.columnKey;
            result.bottomRight.itemKey = endKeyPair.itemKey;
            result.bottomRight.columnKey = startKeyPairs.bottomRight.columnKey;
            break;
        }
        case 'bottomLeft': {
            result.topLeft.itemKey = startKeyPairs.topLeft.itemKey;
            result.topLeft.columnKey = endKeyPair.columnKey;
            result.bottomRight.itemKey = endKeyPair.itemKey;
            result.bottomRight.columnKey = startKeyPairs.bottomRight.columnKey;
            break;
        }
        case 'bottomRight': {
            result.topLeft = startKeyPairs.topLeft;
            result.bottomRight = endKeyPair;
            break;
        }
        case 'left': {
            result.topLeft = endKeyPair;
            result.bottomRight = startKeyPairs.bottomRight;
            break;
        }
        case 'right': {
            result.topLeft = startKeyPairs.topLeft;
            result.bottomRight = endKeyPair;
            break;
        }
    }

    return result;
}

export function getRectParams(startPoint: TPoint, offsetX: number, offsetY: number) {
    const transform = (startPosition: number, offset: number) => {
        if (offset < 0) {
            return {
                size: offset * -1,
                position: startPosition + offset,
            };
        }
        return {
            size: offset,
            position: startPosition,
        };
    };

    const xParams = transform(startPoint.x, offsetX);
    const yParams = transform(startPoint.y, offsetY);

    return {
        x: xParams.position,
        y: yParams.position,
        width: xParams.size,
        height: yParams.size,
    };
}

function getViewportBoundingClientRect(element: HTMLElement): TRect {
    // TODO: Сделать меньше хардкода, но это самый производительный вариант.
    //  Т.к. dynamicGrid крайне специфичный, то пока оставляю так.
    element.classList.remove('tw-contents');
    const containerRect = element.getBoundingClientRect();
    element.classList.add('tw-contents');

    const startFixedPart = element.querySelector(`.${START_FIXED_PART_OBSERVER}`);
    const endFixedPart = element.querySelector(`.${END_FIXED_PART_OBSERVER}`);

    if (startFixedPart) {
        // Можно верить только размерам фиксированных частей.
        // Их координаты будут неверными, т.к. на них висит transform.
        const startFixedRectWidth = startFixedPart.getBoundingClientRect().width;
        containerRect.x += startFixedRectWidth;
        containerRect.width -= startFixedRectWidth;
    }

    if (endFixedPart) {
        // Можно верить только размерам фиксированных частей.
        // Их координаты будут неверными, т.к. на них висит transform.
        containerRect.width -= endFixedPart.getBoundingClientRect().width;
    }

    return containerRect;
}

/**
 * Проверяем, что ячейку можно выделить драгом.
 * Ячейка должна быть полностью видимой.
 * @param e React.MouseEvent.
 */
export function canSelectCellByDrag(e: React.MouseEvent): boolean {
    const cellRect = (e.target as HTMLDivElement).getBoundingClientRect();
    const xSizeToRich = PERCENT_OF_SIZE_TO_SELECT_CELL * cellRect.width;
    const ySizeToRich = PERCENT_OF_SIZE_TO_SELECT_CELL * cellRect.height;
    const availableRect = {
        xStart: cellRect.x + xSizeToRich,
        xEnd: cellRect.x + cellRect.width - xSizeToRich,
        yStart: cellRect.y + ySizeToRich,
        yEnd: cellRect.y + cellRect.height - ySizeToRich,
    };

    return (
        e.pageX >= availableRect.xStart &&
        e.pageX <= availableRect.xEnd &&
        e.pageY >= availableRect.yStart &&
        e.pageY <= availableRect.yEnd
    );
}
