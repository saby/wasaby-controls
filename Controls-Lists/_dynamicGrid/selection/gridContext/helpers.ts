import { TColumnKey, TGridRowSelection, TGridSelection, TPlainSelection } from '../interfaces';

export function resetHasSiblingToDirection(
    selection: TGridRowSelection,
    direction: 'up' | 'down'
): boolean {
    let isChanged = false;

    Object.keys(selection.hasSibling).forEach((key) => {
        if (selection.hasSibling[key][direction]) {
            selection.hasSibling[key][direction] = false;
            isChanged = true;
        }
    });

    return isChanged;
}

function updateSiblingRowSelectionsToDirectionMutable(
    selection: TGridRowSelection,
    siblingSelection: TGridRowSelection,
    currentSelectionKey: TColumnKey,
    direction: 'up' | 'down'
): boolean {
    let isChanged = false;

    const siblingKey = Object.keys(siblingSelection.rowSelection).find(
        (prevKey) =>
            siblingSelection.rowSelection[prevKey].firstColumnKey ===
                selection.rowSelection[currentSelectionKey].firstColumnKey &&
            siblingSelection.rowSelection[prevKey].lastColumnKey ===
                selection.rowSelection[currentSelectionKey].lastColumnKey
    );

    if (typeof siblingKey !== 'undefined') {
        siblingSelection.hasSibling[siblingKey][direction === 'up' ? 'down' : 'up'] = true;
        selection.hasSibling[currentSelectionKey][direction] = true;
        isChanged = true;
    }

    return isChanged;
}

export function updateHasSiblingMutable(
    selection: TGridRowSelection,
    prevSiblingSelection: TGridRowSelection,
    nextSiblingSelection: TGridRowSelection
): [isPrevChanged: boolean, isNextChanged: boolean] {
    Object.keys(selection.rowSelection).forEach((key) => {
        selection.hasSibling[key] = {
            up: false,
            down: false,
        };
    });

    if (!prevSiblingSelection && !nextSiblingSelection) {
        return [false, false];
    }

    let isPrevChanged = false;
    let isNextChanged = false;

    if (prevSiblingSelection) {
        isPrevChanged = resetHasSiblingToDirection(prevSiblingSelection, 'down');
    }

    if (nextSiblingSelection) {
        isNextChanged = resetHasSiblingToDirection(nextSiblingSelection, 'up');
    }

    Object.keys(selection.rowSelection).forEach((currentSelectionKey) => {
        if (prevSiblingSelection) {
            isPrevChanged =
                updateSiblingRowSelectionsToDirectionMutable(
                    selection,
                    prevSiblingSelection,
                    currentSelectionKey as unknown as TColumnKey,
                    'up'
                ) || isPrevChanged;
        }

        if (nextSiblingSelection) {
            isNextChanged =
                updateSiblingRowSelectionsToDirectionMutable(
                    selection,
                    nextSiblingSelection,
                    currentSelectionKey as unknown as TColumnKey,
                    'down'
                ) || isNextChanged;
        }
    });

    if (isPrevChanged) {
        prevSiblingSelection.hasSibling = { ...prevSiblingSelection.hasSibling };
    }

    if (isNextChanged) {
        nextSiblingSelection.hasSibling = { ...nextSiblingSelection.hasSibling };
    }

    return [isPrevChanged, isNextChanged];
}

export function isEqualPlainSelections(
    selection1: TPlainSelection,
    selection2: TPlainSelection
): boolean {
    return compareKeys(Object.keys(selection1), Object.keys(selection2), (a, b) => {
        return compareKeys(selection1[a], selection2[b]);
    });
}

function compareKeys<T>(
    keys1: T[],
    keys2: T[],
    extraComparator?: (key1: T, key2: T) => boolean
): boolean {
    if (keys1.length !== keys2.length) {
        return false;
    } else if (keys1.length === 0) {
        return true;
    } else {
        const sorted1 = keys1.sort();
        const sorted2 = keys2.sort();

        // Ключи есть, их одинаковое количество и они отсортированы.
        for (let i = 0; i < sorted1.length; i++) {
            if (sorted1[i] !== sorted2[i]) {
                return false;
            } else if (extraComparator && !extraComparator(sorted1[i], sorted2[i])) {
                return false;
            }
        }

        return true;
    }
}

export function convertGridSelectionToPlainRecord(selection: TGridSelection): TPlainSelection {
    return Object.keys(selection).reduce((acc, sKey) => {
        return {
            ...acc,
            [sKey]: Object.keys(selection[sKey].rowSelection).reduce(
                (rowSelectedKeys, rsKey) => [
                    ...rowSelectedKeys,
                    ...selection[sKey].rowSelection[rsKey].selectedKeys,
                ],
                []
            ),
        };
    }, {});
}

export function convertPlainSelectionToGridSelection(selection: TPlainSelection): TGridSelection {
    const gridSelection: TGridSelection = {};
    // TODO: Не реализовано, будет делаться тут, если это вообще требуется.
    // Сейчас смена набора выбранных ячеек прикладным разработчиком через слайс не работает.
    return gridSelection;
}
