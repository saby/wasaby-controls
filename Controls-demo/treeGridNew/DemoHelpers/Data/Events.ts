import { IColumn } from 'Controls/grid';
import { IData } from '../Interface';

export const Events = {
    getColumns: (): IColumn[] => {
        return [
            {
                displayProperty: 'title',
                width: '',
            },
        ];
    },
    getDataBeforeExpanded: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед разворотом узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataBeforeCollapsed: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Асинхронная обработка перед сворачиванием узла',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataCollapsedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно свернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
    getDataExpandedItemsChanged: (): IData[] => {
        return [
            {
                key: 1,
                title: 'Этот узел нельзя развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 11, title: 'Лист 1-1', parent: 1, nodeType: null },
            { key: 12, title: 'Лист 1-2', parent: 1, nodeType: null },
            {
                key: 2,
                title: 'Этот узел можно развернуть',
                parent: null,
                nodeType: true,
            },
            { key: 21, title: 'Лист 2-1', parent: 2, nodeType: null },
            { key: 22, title: 'Лист 2-2', parent: 2, nodeType: null },
        ];
    },
};
