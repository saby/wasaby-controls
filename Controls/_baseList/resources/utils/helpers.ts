/**
 * @kaizen_zone b9244594-2ac8-4db9-8c67-cd873d12a1c4
 */
import { Model } from 'Types/entity';
import { CollectionItem } from 'Controls/display';
import { CrudEntityKey } from 'Types/source';

/**
 * TODO нужно выпилить этот метод при переписывании моделей. item.getContents() должен возвращать Record
 *  https://online.sbis.ru/opendoc.html?guid=acd18e5d-3250-4e5d-87ba-96b937d8df13
 * @param item
 */
export function getPlainItemContents(item: CollectionItem<Model>): Model {
    let contents = item.getContents();
    if (item['[Controls/_baseTree/BreadcrumbsItem]'] || item.breadCrumbs) {
        contents = contents[(contents as any).length - 1];
    }
    return contents;
}

export function getValidatedItemKey(key: string): CrudEntityKey {
    // Если в начале строки точка, то нельзя приводить этот ключ к числу, т.к. получим число 0.XXXX
    return isNaN(key) || (typeof key === 'string' && key[0] === '.')
        ? key
        : Number(key);
}

export function getKey(item: Model | string): CrudEntityKey {
    if (item['[Types/_entity/Model]']) {
        return item.getKey();
    }

    return item as string;
}
