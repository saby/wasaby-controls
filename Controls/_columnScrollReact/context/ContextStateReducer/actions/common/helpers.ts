import { isEdgeAnimated } from '../../../../common/helpers';
import { DebugLogger } from '../../../../common/DebugLogger';
import { EdgeState } from '../../../../common/types';
import { TState } from '../../../ContextStateReducer/TState';

/**
 * Состояние триггеров/границ области скроллирования.
 * @private
 */
interface IEdgesState {
    /**
     * Состояние левого триггера/левой границы области скроллирования.
     */
    left: EdgeState;
    /**
     * Состояние правого триггера/правой границы области скроллирования.
     */
    right: EdgeState;
}

/**
 * Метод мутации состояния границ области скроллирования на контексте.
 * @param {TState} oldState Старое состояние контекста.
 * @param {TState} newState Новое состояние контекста.
 * @param {IEdgesState} edges Новое состояние границ области скроллирования.
 */
export function updateEdgesStateMutable(oldState: TState, newState: TState, edges: IEdgesState) {
    if (oldState.leftEdgeState !== edges.left) {
        newState.leftEdgeState = edges.left;
    }
    if (oldState.rightEdgeState !== edges.right) {
        newState.rightEdgeState = edges.right;
    }
}

/**
 * Метод мутации состояния тригеров на контексте.
 * @param {TState} oldState Старое состояние контекста.
 * @param {TState} newState Новое состояние контекста.
 * @param {IEdgesState} triggers Новое состояние треггеров.
 */
export function updateTriggersStateMutable(
    oldState: TState,
    newState: TState,
    triggers: IEdgesState
): boolean {
    let isChanged = false;

    if (oldState.leftTriggerEdgeState !== triggers.left && !isEdgeAnimated(triggers.left)) {
        isChanged = true;
        newState.leftTriggerEdgeState = triggers.left;
    }

    if (oldState.rightTriggerEdgeState !== triggers.right && !isEdgeAnimated(triggers.right)) {
        isChanged = true;
        newState.rightTriggerEdgeState = triggers.right;
    }

    if (isChanged) {
        DebugLogger.bordersStateChanged(
            {
                left: oldState.leftTriggerEdgeState,
                right: oldState.rightTriggerEdgeState,
            },
            triggers,
            'triggers'
        );
    }

    return isChanged;
}
