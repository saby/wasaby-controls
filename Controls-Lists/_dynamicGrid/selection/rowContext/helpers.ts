import { TColumnKey, ISelection, TRowSelection, TColumns } from '../interfaces';
import { datesEqualByQuantum, TQuantumType } from '../../render/utils';

export function getColumnIndexByKey(
    columnKey: TColumnKey,
    columns: TColumns,
    quantum: TQuantumType
): number {
    const date = new Date(columnKey);
    return columns.findIndex((d) => datesEqualByQuantum(date, d, quantum));
}

export function createNewSelection(newKeys: TColumnKey[]): ISelection {
    return {
        rootKey: newKeys[0],
        firstColumnKey: newKeys[0],
        lastColumnKey: newKeys[newKeys.length - 1],
        selectedKeys: newKeys,
        hasBorderTop: true,
        hasBorderBottom: true,
        gridColumnStart: undefined,
        gridColumnEnd: undefined,
    };
}

export function unshiftFromSelectionMutable(
    rowSelection: TRowSelection,
    selection: ISelection,
    columnsToRoot: Record<TColumnKey, TColumnKey>
): ISelection {
    const columnKey = selection.firstColumnKey;
    const rightSelection: ISelection = createNewSelection(selection.selectedKeys.slice(1));

    rightSelection.selectedKeys.forEach((k) => {
        columnsToRoot[k] = rightSelection.rootKey;
    });

    delete rowSelection[columnsToRoot[columnKey]];
    delete columnsToRoot[columnKey];

    rowSelection[rightSelection.rootKey] = rightSelection;

    return rightSelection;
}

export function popFromSelectionMutable(
    rowSelection: TRowSelection,
    selection: ISelection,
    columnsToRoot: Record<TColumnKey, TColumnKey>
): ISelection {
    const columnKey = selection.lastColumnKey;
    const leftSelection: ISelection = createNewSelection(
        selection.selectedKeys.slice(0, selection.selectedKeys.length - 1)
    );

    leftSelection.selectedKeys.forEach((k) => {
        columnsToRoot[k] = leftSelection.rootKey;
    });

    delete rowSelection[columnsToRoot[columnKey]];
    delete columnsToRoot[columnKey];
    rowSelection[leftSelection.rootKey] = leftSelection;

    return leftSelection;
}

export function splitSelectionMutable(
    rowSelection: TRowSelection,
    selection: ISelection,
    columnsToRoot: Record<TColumnKey, TColumnKey>,
    columnKey: TColumnKey
): [leftSelection: ISelection, rightSelection: ISelection] {
    const leftSelection = createNewSelection(
        selection.selectedKeys.slice(0, selection.selectedKeys.indexOf(columnKey))
    );

    leftSelection.selectedKeys.forEach((k) => {
        columnsToRoot[k] = leftSelection.rootKey;
    });

    const rightSelection = createNewSelection(
        selection.selectedKeys.slice(selection.selectedKeys.indexOf(columnKey) + 1)
    );

    rightSelection.selectedKeys.forEach((k) => {
        columnsToRoot[k] = rightSelection.rootKey;
    });

    delete rowSelection[columnsToRoot[columnKey]];
    delete columnsToRoot[columnKey];

    rowSelection[leftSelection.rootKey] = leftSelection;
    rowSelection[rightSelection.rootKey] = rightSelection;

    return [leftSelection, rightSelection];
}

export function deleteSelectionMutable(
    rowSelection: TRowSelection,
    selection: ISelection,
    columnsToRoot: Record<TColumnKey, TColumnKey>
): void {
    const columnKey = selection.rootKey;
    delete rowSelection[columnsToRoot[columnKey]];
    delete columnsToRoot[columnKey];
}

export function handleAddToSelectionMutable(
    rowSelection: TRowSelection,
    columnsToRoot: Record<TColumnKey, TColumnKey>,
    columnKey: TColumnKey,
    prevColumnKey: TColumnKey,
    nextColumnKey: TColumnKey
): ISelection {
    let joinSelection: ISelection;

    if (columnsToRoot[prevColumnKey]) {
        joinSelection = rowSelection[columnsToRoot[prevColumnKey]];

        joinSelection.selectedKeys.push(columnKey);
        joinSelection.lastColumnKey = columnKey;
        columnsToRoot[columnKey] = columnsToRoot[prevColumnKey];
    }

    if (columnsToRoot[nextColumnKey]) {
        const nextSelection = rowSelection[columnsToRoot[nextColumnKey]];
        const nextSelectionRoot = columnsToRoot[nextColumnKey];

        if (joinSelection) {
            joinSelection.selectedKeys.push(...nextSelection.selectedKeys);
            joinSelection.lastColumnKey = nextSelection.lastColumnKey;
            nextSelection.selectedKeys.forEach((key) => {
                columnsToRoot[key] = joinSelection.rootKey;
            });
            delete rowSelection[nextSelectionRoot];
        } else {
            joinSelection = nextSelection;

            joinSelection.selectedKeys.unshift(columnKey);
            joinSelection.firstColumnKey = columnKey;
            columnsToRoot[columnKey] = columnsToRoot[nextColumnKey];
        }
    }

    if (!joinSelection) {
        joinSelection = createNewSelection([columnKey]);
        rowSelection[columnKey] = joinSelection;
        columnsToRoot[columnKey] = columnKey;
    }

    return joinSelection;
}

export function getGridColumnCSSStyle(selection: ISelection, columns: TColumns, quantum: TQuantumType): {
    gridColumnStart: number;
    gridColumnEnd: number;
} {
    let gridColumnStart = getColumnIndexByKey(selection.firstColumnKey, columns, quantum);
    let gridColumnEnd = getColumnIndexByKey(selection.lastColumnKey, columns, quantum);

    if (gridColumnStart === -1 || gridColumnEnd === -1) {
        if (gridColumnStart === -1 && gridColumnEnd === -1) {
            gridColumnStart = undefined;
            gridColumnEnd = undefined;
        } else if (gridColumnStart === -1) {
            gridColumnStart = 1;
        } else if (gridColumnEnd === -1) {
            gridColumnEnd = -1;
        }
    } else {
        gridColumnStart = gridColumnStart + 1;
        gridColumnEnd = gridColumnEnd + 1 + 1;
    }

    return {gridColumnStart, gridColumnEnd};
}
