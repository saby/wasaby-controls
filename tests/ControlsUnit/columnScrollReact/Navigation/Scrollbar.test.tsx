/**
 * @jest-environment jsdom
 */
jest.mock('Controls/_columnScrollReact/Navigation/PlatformScrollbar/ScrollBarBase', () => {
    return {
        __esModule: true,
        default: (props) => {
            return (
                <div data-qa="mocked_ScrollbarComponent">
                    <div data-qa="mocked_ScrollbarComponent_props">
                        {JSON.stringify(props, null, 4)}
                    </div>
                </div>
            );
        },
    };
});

import { unmountComponentAtNode } from 'react-dom';
import { WasabyEvents } from 'UICore/Events';

import { renderAnyComponent } from '../forTestsOnly/helpers';

import {
    ScrollbarNavigationComponent,
    IScrollbarNavigationComponentProps,
} from 'Controls/_columnScrollReact/Navigation/Scrollbar';

import { QA_SELECTORS } from 'Controls/columnScrollReact';

describe('ControlsUnit/columnScrollReact/Navigation/Scrollbar', () => {
    let container = null;

    beforeEach(() => {
        container = document.createElement('div');
        WasabyEvents.initInstance(container);
        document.body.appendChild(container);
    });

    afterEach(() => {
        unmountComponentAtNode(container);
        WasabyEvents.destroyInstance(container);
        container.remove();
        container = null;
    });

    it('На контрол можно навесить свой класс', () => {
        const props: IScrollbarNavigationComponentProps = {
            startFixedWidth: 100,
            endFixedWidth: 0,
            contentWidth: 1000,
            viewPortWidth: 500,
            position: 0,
        };
        const { getByQA } = renderAnyComponent(
            <ScrollbarNavigationComponent {...props} className="testClassName__scrollBar" />,
            { container }
        );

        const scrollBarContainer = getByQA(QA_SELECTORS.NAVIGATION_SCROLLBAR) as HTMLDivElement;

        expect(
            scrollBarContainer.className.indexOf('testClassName__scrollBar') !== -1
        ).toBeTruthy();
    });

    describe('Платформенному скроллбару должны проксироваться опции.', () => {
        const testCases: IScrollbarNavigationComponentProps[] = [
            // Только обязательные
            {
                startFixedWidth: 100,
                endFixedWidth: 0,
                contentWidth: 1000,
                viewPortWidth: 500,
                position: 0,
            },
            // Полные
            {
                startFixedWidth: 200,
                endFixedWidth: 0,
                contentWidth: 900,
                viewPortWidth: 400,
                position: 20,
                className: 'testClassName__1__scrollBar',
                readOnly: true,
                scrollBarValign: 'bottom',
            },
        ];

        testCases.forEach((props, index) => {
            it(`case #${index}.`, () => {
                const { asFragment } = renderAnyComponent(
                    <ScrollbarNavigationComponent {...props} />,
                    { container }
                );
                expect(asFragment()).toMatchSnapshot();
            });
        });
    });
});
