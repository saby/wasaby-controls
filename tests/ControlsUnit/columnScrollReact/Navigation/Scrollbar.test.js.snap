// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlsUnit/columnScrollReact/Navigation/Scrollbar Платформенному скроллбару должны проксироваться опции. case #0. 1`] = `
<DocumentFragment>
  <div
    class="tw-w-full controls-ColumnScrollReact__acceleration "
    data-qa="ColumnScrollReact__navigation__scrollbar"
  >
    <div
      data-qa="mocked_ScrollbarComponent"
    >
      <div
        data-qa="mocked_ScrollbarComponent_props"
      >
        {
    "style": {
        "top": 0
    },
    "position": 0,
    "contentSize": 900,
    "direction": "horizontal",
    "thumbSize": "s",
    "thumbStyle": "unaccented",
    "trackVisible": true,
    "hasStartPadding": true,
    "hasEndPadding": true,
    "shouldSetMarginTop": true
}
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ControlsUnit/columnScrollReact/Navigation/Scrollbar Платформенному скроллбару должны проксироваться опции. case #1. 1`] = `
<DocumentFragment>
  <div
    class="tw-w-full controls-ColumnScrollReact__acceleration testClassName__1__scrollBar"
    data-qa="ColumnScrollReact__navigation__scrollbar"
  >
    <div
      data-qa="mocked_ScrollbarComponent"
    >
      <div
        data-qa="mocked_ScrollbarComponent_props"
      >
        {
    "style": {
        "top": 0
    },
    "readOnly": true,
    "position": 20,
    "contentSize": 700,
    "direction": "horizontal",
    "thumbSize": "s",
    "thumbStyle": "unaccented",
    "trackVisible": true,
    "hasStartPadding": true,
    "hasEndPadding": true,
    "shouldSetMarginTop": false
}
      </div>
    </div>
  </div>
</DocumentFragment>
`;
